### PR TITLE
query frontend, query UI: Native histogram support

### PR DESCRIPTION
**Pull request by rabenhorst**
As followup to #6032 , this PR adds native histogram support for query frontend and query UI. Additionally this PR adds native histogram support in the `ApplyCounterResetsSeriesIterator`, which is required for histogram rates.

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

- Native histogram support for query frontend
- Native histogram support for query UI (table and graph)
- Native histogram support in `ApplyCounterResetsSeriesIterator`
- [Enable native histograms rates](https://github.com/thanos-io/thanos/pull/6071/commits/d6b359ea9d49ef972ca276403d1e37cd5f63f3c4).

## Verification

Unit tests and e2e tests.

### CODE REVIEW

File: internal/cortex/querier/queryrange/queryrange.proto [hunk 1]
```diff
@@ -98,18 +98,39 @@ message PrometheusResponseQueryableSamplesStatsPerStep {
 message SampleStream {
   repeated cortexpb.LabelPair labels = 1 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "metric", (gogoproto.customtype) = "github.com/thanos-io/thanos/internal/cortex/cortexpb.LabelAdapter"];
   repeated cortexpb.Sample samples = 2 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "values"];
+  repeated SampleHistogramPair histograms = 3 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "histogram"];
 }
 
 message Sample {
   repeated cortexpb.LabelPair labels = 1 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "metric", (gogoproto.customtype) = "github.com/thanos-io/thanos/internal/cortex/cortexpb.LabelAdapter"];
-  cortexpb.Sample sample = 2 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "value"];
+  double sampleValue = 2 [(gogoproto.jsontag) = "value"];
```
Comment (rabenhorst) [inline internal/cortex/querier/queryrange/queryrange.proto:106]: I understand that this is a breaking change and invalidating cache is necessary. I made the change to be closer to the prometheus common model.

Comment (fpetkovski) [reply internal/cortex/querier/queryrange/queryrange.proto:106]: I think this is fine as long as it's reflected in the CHANGELOG.

**Review by fpetkovski — DISMISSED** [review]

File: internal/cortex/querier/queryrange/queryrange.proto [hunk 2]
```diff
@@ -98,18 +98,39 @@ message PrometheusResponseQueryableSamplesStatsPerStep {
 message SampleStream {
   repeated cortexpb.LabelPair labels = 1 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "metric", (gogoproto.customtype) = "github.com/thanos-io/thanos/internal/cortex/cortexpb.LabelAdapter"];
   repeated cortexpb.Sample samples = 2 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "values"];
+  repeated SampleHistogramPair histograms = 3 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "histogram"];
```
Comment (yeya24) [inline internal/cortex/querier/queryrange/queryrange.proto:101]: I am not very familiar with the new histogram, so for a sample stream, we will only have either samples or histograms? Can we have them both the same time?

File: CHANGELOG.md [hunk 3]
```diff
@@ -38,6 +38,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 - [#6010](https://github.com/thanos-io/thanos/pull/6010) *: Upgrade Prometheus to v0.41.0.
 - [#5999](https://github.com/thanos-io/thanos/pull/5999) *: Upgrade Alertmanager dependency to v0.25.0.
 - [#5887](https://github.com/thanos-io/thanos/pull/5887) Tracing: Make sure rate limiting sampler is the default, as was the case in version pre-0.29.0.
+- [#6071](https://github.com/thanos-io/thanos/pull/6071) Query Frontend: *breaking :warning:* Aligned with Prometheus common model (cache reset required)
```
Comment (yeya24) [inline CHANGELOG.md:41]: Can we use a more descriptive sentence about this change. Does this mean QFE can return native histograms in its response?
Btw since v0.31.0 rc is there let's update the entry location.

File: pkg/compact/downsample/downsample.go [hunk 4]
```diff
@@ -617,15 +618,21 @@ func (it *ApplyCounterResetsSeriesIterator) Next() chunkenc.ValueType {
 		if it.i >= len(it.chks) {
 			return chunkenc.ValNone
 		}
-		if it.chks[it.i].Next() == chunkenc.ValNone {
+		it.lastValType = it.chks[it.i].Next()
```
Comment (yeya24) [inline pkg/compact/downsample/downsample.go:621]: Is this change related to this pr? Downsample/Compact side change seems not related to QFE and UI.

Comment (rabenhorst) [reply pkg/compact/downsample/downsample.go:633]: You are right. It's not directly related but it's a fix so that `rate()` works for native histograms: https://github.com/thanos-io/thanos/blob/fdeea3917591fc363a329cbe23af37c6fff0b5f0/pkg/query/iter.go#L131

I will add this information to the first comment.

File: CHANGELOG.md (hunk 3 above)
Comment (rabenhorst) [reply CHANGELOG.md]: Done.

File: internal/cortex/querier/queryrange/queryrange.proto (hunk 2 above)
Comment (rabenhorst) [reply internal/cortex/querier/queryrange/queryrange.proto:101]: From my understanding and according to the Prometheus common code a SampleStream can have both samples and histograms: https://github.com/prometheus/common/blob/4827ac3e6ff608c6959b8a53665e715904297a69/model/value.go#L195-L226

Comment (fpetkovski) [reply internal/cortex/querier/queryrange/queryrange.proto:101]: Correct, it is possible for a PromQL query to return both histograms and floats. Something like `rate` will do this because it is applied independently on each series.

**Comment by GiedriusS** [comment]
Sorry for the delay. Is this ready for review?

**Comment by rabenhorst** [comment]
> Sorry for the delay. Is this ready for review?

Yes it's ready and I just rebased it.

File: internal/cortex/querier/queryrange/query_range.go [hunk 5]
```diff
@@ -657,7 +690,24 @@ func matrixMerge(resps []*PrometheusResponse) []SampleStream {
...
+			if len(existing.Histograms) > 0 && len(stream.Histograms) > 0 {
+				existingEndTs := existing.Histograms[len(existing.Histograms)-1].GetTimestamp()
+				if existingEndTs == stream.Histograms[0].GetTimestamp() {
+					stream.Histograms = stream.Histograms[1:]
+				} else if existingEndTs > stream.Histograms[0].GetTimestamp() {
+					stream.Histograms = SliceHistogram(stream.Histograms, existingEndTs)
+				}
+			}
+
+			if len(stream.Samples) > 0 {
```
Comment (GiedriusS) [inline internal/cortex/querier/queryrange/query_range.go:718]: Do we need these `if`s? Maybe we can just call append instead? :thinking:

File: internal/cortex/querier/queryrange/results_cache_test.go [hunk 6]
```diff
@@ -36,6 +37,14 @@ var (
 		Query: "sum(container_memory_rss) by (namespace)",
 		Stats: "all",
 	}
+	parsedHistogramRequest = &PrometheusRequest{
+		Path:  "/api/v1/query_range",
+		Start: 1536673680 * 1e3,
+		End:   1536716898 * 1e3,
+		Step:  120 * 1e3,
+		Query: "fake_histogram",
```
Comment (GiedriusS) [inline internal/cortex/querier/queryrange/results_cache_test.go:45]: Nit: if the metric is `{"fake": "histogram"}` then perhaps this query should be `{fake="histogram"}`? This doesn't change anything, just to be correct :thinking:

File: test/e2e/native_histograms_test.go [hunk 7]
```diff
@@ -96,48 +110,263 @@ func TestWriteNativeHistograms(t *testing.T) {
...
+	queryAndAssertSeries(t, ctx, querier.Endpoint("http"), func() string { return testHistogramMetricName }, time.Now, promclient.QueryOptions{Deduplicate: true}, []model.Metric{
+		{
+			"__name__":   testHistogramMetricName,
+			"prometheus": "prom-ha",
+			"foo":        "bar",
+		},
+	})
+
+	vals, err := querier.SumMetrics([]string{"http_requests_total"})
+	e2emon.WithLabelMatchers(matchers.MustNewMatcher(matchers.MatchEqual, "handler", "query"))
```
Comment (GiedriusS) [inline test/e2e/native_histograms_test.go:183]: This value is unused? :thinking:

Comment (rabenhorst) [reply test/e2e/native_histograms_test.go]: You are tight, should be argument and I fixed it.

File: pkg/queryfrontend/downsampled.go [hunk 8]
```diff
@@ -97,14 +98,23 @@ forLoop:
...
-	for _, sampleStream := range res[1:] {
-		if ts := sampleStream.Samples[0].TimestampMs; ts < minTs {
+
+	minTs := int64(math.MaxInt64)
+
+	for _, sampleStream := range res {
+		var ts int64
+		if len(sampleStream.Samples) > 0 {
+			ts = sampleStream.Samples[0].TimestampMs
+		} else if len(sampleStream.Histograms) > 0 {
```
[the remark this answers is not in the document]
Comment (rabenhorst) [reply pkg/queryfrontend/downsampled.go]: Fixed that and added test cases for mix sampleStream.

File: pkg/queryfrontend/downsampled.go [hunk 9]
```diff
@@ -97,14 +98,23 @@ forLoop:
 // Each SampleStream within r.Data.Result must be sorted by timestamp.
 func minResponseTime(r queryrange.Response) int64 {
 	var res = r.(*queryrange.PrometheusResponse).Data.Result
-	if len(res) == 0 || len(res[0].Samples) == 0 {
+	if len(res) == 0 || len(res[0].Samples) == 0 && len(res[0].Histograms) == 0 {
```
[the remark this answers is not in the document]
Comment (rabenhorst) [reply pkg/queryfrontend/downsampled.go]: Done.

**Review by GiedriusS — COMMENTED** [review]
Please avoid rewriting history in the future with `--force` because now I'm forced to review thousands of lines again. :cry:

**Comment by rabenhorst** [comment]
> Please avoid rewriting history in the future with `--force` because now I'm forced to review thousands of lines again. 😢

Sry for that, will avoid it in the future. I realized what I have done when it was too late already. 😞

File: test/e2e/native_histograms_test.go [hunk 10]
```diff
@@ -38,35 +45,53 @@ func TestQueryNativeHistograms(t *testing.T) {
...
 	})
 
-	// Query histogram using group function to test pushdown.
-	queryAndAssert(t, ctx, querier.Endpoint("http"), func() string { return "group(fake_histogram)" }, time.Now, promclient.QueryOptions{Deduplicate: true}, model.Vector{
-		&model.Sample{
-			Value:  1,
-			Metric: model.Metric{},
-		},
+	t.Run("query histogram using group function for testing pushdown", func(t *testing.T) {
+		queryAndAssert(t, ctx, querier.Endpoint("http"), func() string { return fmt.Sprintf("group(%v)", testHistogramMetricName) }, ts, promclient.QueryOptions{Deduplicate: true}, model.Vector{
```
Comment (GiedriusS) [inline test/e2e/native_histograms_test.go:81]: I think we forgot to enable the pushdown feature flag in `querier` :smile:

File: pkg/compact/downsample/downsample.go [hunk 11]
```diff
@@ -674,11 +681,11 @@ func (it *ApplyCounterResetsSeriesIterator) At() (t int64, v float64) {
 
 // TODO(rabenhorst): Needs to be implemented for native histogram support.
```
Comment (GiedriusS) [inline pkg/compact/downsample/downsample.go:682]: Shall we remove this?

File: pkg/ui/react-app/src/pages/graph/Graph.tsx [hunk 12]
```diff
@@ -16,7 +16,7 @@ require('jquery.flot.tooltip');
 export interface GraphProps {
   data: {
     resultType: string;
-    result: Array<{ metric: Metric; values: [number, string][] }>;
+    result: Array<{ metric: Metric; values: [number, string][]; histograms?: [number, Histogram][] }>;
```
Comment (GiedriusS) [inline pkg/ui/react-app/src/pages/graph/Graph.tsx:19]: Let's use `SampleValue` and `SampleHistogram` types here? :thinking:

File: pkg/compact/downsample/downsample_test.go [hunk 13]
```diff
@@ -799,6 +799,62 @@ func TestApplyCounterResetsIterator(t *testing.T) {
...
+			var res []*histogramPair
+			for x.Next() != chunkenc.ValNone {
+				t, h := x.AtHistogram()
+				res = append(res, &histogramPair{t, h})
+			}
+			testutil.Ok(t, x.Err())
+			testutil.Equals(t, tcase.expected, res)
+
+			for i := range res[1:] {
+				testutil.Assert(t, res[i+1].t >= res[i].t, "sample time %v is not monotonically increasing. previous sample %v is older", res[i+1], res[i])
```
[the remark this answers is not in the document]
Comment (GiedriusS) [reply pkg/compact/downsample/downsample_test.go:851]: Ahh, I see, somehow missed that `i` starts from `0` and it will not go until `39` :+1: this is a nice trick.

File: test/e2e/native_histograms_test.go (hunk 10 above)
Comment (rabenhorst) [reply test/e2e/native_histograms_test.go:83]: You are right. I enabled it for this querier.

**Review by GiedriusS — COMMENTED** [review]
I think Cortex will be able to use this PR as an inspiration to implement native histograms on their side, right? (https://github.com/cortexproject/cortex/issues/5060)

**Review by GiedriusS — APPROVED** [review]
LGTM minus those small nits, I think you forgot to add DCO on your latest commit :smile:

**Comment by rabenhorst** [comment]
> LGTM minus those small nits, I think you forgot to add DCO on your latest commit 😄

I addressed the comments and fixe DCO. 

Btw. I track progress for native histograms in https://github.com/thanos-io/thanos/issues/5907.
